### ISSUE TITLE
Support readline or equivalent alternatives

### ISSUE THREAD

**Issue by achimnol** [vxgmichel/aioconsole#7]
Googling leads me to [this SO question](https://stackoverflow.com/questions/45659928/readline-python-module-does-not-work-with-aioconsole-ainput) but there is no reply yet.

Python already provides a binding module to GNU readline or libedit, but this does not work together with aioconsole since it does not have coroutine-based interfaces. Maybe we need a new lower-level wrapper for the readline library.

This issue is both question and feature request:
 * Is there any plan to support readline or readline-style interactive prompts?
 * Are there any known/fundamental technical issues related to supporting readline (except that the stdlib readline does not have coroutines)?
   - Could we just make the user prompt to be synchronous while leaving user codes running between prompts asynchronous?

**Comment by dufferzafar** [comment]
I don't know much about readline, but will supporting it allow us to use left/right arrow keys in the prompts?

**Comment by vxgmichel** [comment]
@achimnol Thank you for creating an issue, this is very interesting.

I'm not knowledgeable enough about `readline` to answer all your questions, I'll have to start an investigation. What I can tell you though is that `aioconsole` aims at providing an abstract asynchronous interface to arbitrary `stdin/stdout`, with methods similar to the [Stream API](https://docs.python.org/3/library/asyncio-stream.html). 

Meaning those streams can either come from:
- the console (standard unix streams)
- an application like IDLE  (or Windows, FWIW)
- a TCP socket (for serving python prompts over the network)

As @dufferzafar pointed out, it would nice to have support of left/right arrow keys for all those use cases, but I have no idea whether this is doable yet. It would also be worth looking into [prompt_toolkit](https://github.com/jonathanslenders/python-prompt-toolkit) as well, since it has asyncio support.

Any help is more than welcome by the way :wink:

**Comment by dufferzafar** [comment]
Yeah, I recently had to build an asyncio based CLI and found out that prompt_toolkit has support for it. I think that is our best bet to create an awesome async-cli.

Haven't looked into the specifics yet, but will definitely post anything I learn here.

**Comment by lilydjwg** [comment]
I tried to add support for readline with low-level, callback-based readline API, but it segfaults mysteriously in the callback. I might investigate further, but this will surely only support read from terminal.

If you connect via TCP, you can just use `socat readline tcp:host:port` to have basic readline support.

**Comment by TD22057** [comment]
@lilydjwg How does the socat option work?  I ran the echo example and I can connect fine w/ nc but when I try to say socat readline anything, I just get an error (using the latest version of socat):
```
socat readline tcp:127.0.0.1:9999 
2017/10/27 11:21:45 socat[603] E unknown device/address "readline"
```

**Comment by vxgmichel** [comment]
@TD22057 There seems to be a problem with [socat and readline](https://softwarerecs.stackexchange.com/a/37694/34919). 

Instead you can use [rlwrap](https://github.com/hanslub42/rlwrap):
```bash
# Works with local apython console
$ rlwrap apython

# Also works with a remote console
$ apython --serve locahost:8888
The console is being served on 127.0.0.1:8888

# with both netcat and socat
$ rlwrap nc localhost 8888
$ rlwrap socat - tcp:localhost 8888
```
I actually didn't know about rlwrap but it seems to fix the issue. It seems to be a valid solution for remote consoles, and an acceptable workaround for local consoles. I'll update the documentation.

Thanks for your post.

**Comment by TD22057** [comment]
rlwrap works perfectly and is exactly what I need for my application.  Thanks for this very cool project.

**Comment by vxgmichel** [comment]
Hi all,

I just implemented [readline support (PR #9)](https://github.com/vxgmichel/aioconsole/pull/9) for local console through the `apython` script. It seems to work fine and it even provides some level of auto-completion (only for the built-ins).

The implementation is hackish but contained, and can even be disabled using the `--no-readline` argument. It works by starting the actual program in a subprocess and controlling the standard streams in order to take advantage of the readline support provided by the `input` built-in.

Feel free to review the PR if you're interested. Otherwise, I'll probably merge it next week (or not, if I change my mind and finally consider it's too hackish :) ).

Thanks for your inputs!
